ARUHA-882 Fix Indeterministic tests

antban opened the pull request:
[ARUHA-882: Indeterministic tests]

## Description
Right now tests sometimes passing and depends on the order of runs, making it impossible to run tests in parallel.

## Deployment Notes
No additional notes.

v-stepanov commented on src/acceptance-test/java/org/zalando/nakadi/repository/db/SubscriptionDbRepositoryTest.java:
```diff
@@ -91,88 +86,87 @@ public void whenGetSubscriptionByIdThenOk() throws Exception {
     }
 
     @Test
-    public void whenListSubscriptionsThenOk() throws ServiceUnavailableException {
```
was it replaced by something else?

antban replied:
No, it is not replaced, cause other tests present there as well which are testing this kind of call

v-stepanov commented on src/acceptance-test/java/org/zalando/nakadi/webservice/BaseAT.java:
```diff
@@ -46,26 +36,21 @@
-            storage.setType(Storage.Type.KAFKA);
-            storage.setConfiguration(new Storage.KafkaConfiguration(null, null, ZOOKEEPER_URL, ""));
             STORAGE_DB_REPOSITORY.createStorage(storage);
+        } catch (final Exception ignore) {
```
Which exception do you ignore here? Is it DuplicatedStorageException ?

antban replied:
@v-stepanov Probably yes. will change Exception to DuplicatedStorageException

v-stepanov commented on src/acceptance-test/java/org/zalando/nakadi/webservice/EventStreamReadingAT.java:
```diff
@@ -315,20 +326,22 @@ public void whenInvalidCursorsThenPreconditionFailed() {
     }
 
-    @Test(timeout = 10000)
+    @Test(timeout = 30000)
```
Does this test really need 30 seconds as a timeout? I think that the value of this test is not enough for spending up to 30 seconds on it. Also I think it can be replaced by Unit Test.

antban replied:
Probably it was an investigation of problem, and I forgot to change it back. Will change it back.

v-stepanov commented on src/acceptance-test/java/org/zalando/nakadi/webservice/EventStreamReadingAT.java:
```diff
@@ -416,29 +429,30 @@ private int countStatusCode(final List<CompletableFuture<HttpURLConnection>> fut
     }
 
-    @Test(timeout = 10000)
+    @Test(timeout = 60000)
```
Just like a previous comment: I think we can replace this with unit test. I don't want to spend up to 60 seconds here.

antban replied:
Probably it was an investigation of problem, and I forgot to change it back. Will change it back.

v-stepanov commented:
The score is 5:5

v-stepanov commented on src/acceptance-test/java/org/zalando/nakadi/webservice/SettingsControllerAT.java:
```diff
@@ -2,46 +2,37 @@
+            Assert.assertNotNull(CURATOR.checkExists()
+                    .forPath("/nakadi/blacklist/consumers/event_types/" + eventType.getName()));
+        } finally {
+            whitelist(eventType.getName(), BlacklistService.Type.CONSUMER_ET);
```
why waste time whitelisting it?

v-stepanov commented on src/acceptance-test/java/org/zalando/nakadi/webservice/SettingsControllerAT.java:
```diff
@@ -59,38 +50,17 @@ public void testUnblockFlooder() throws Exception {
+                            .body("consumers.event_types", hasItems(eventType.getName())),
+                    1000, 200);
+        } finally {
+            whitelist(eventType.getName(), BlacklistService.Type.CONSUMER_ET);
```
why do you want to whitelist it at the end?

antban replied:
@v-stepanov As a general rule with shared data if i created it, then I will delete it

v-stepanov replied:
I thought we have a "dirty DB" concept for tests now.

v-stepanov commented on src/acceptance-test/java/org/zalando/nakadi/webservice/timelines/TimelineConsumptionTest.java:
```diff
@@ -34,16 +33,17 @@ public static void setupEventTypeWithEvents() throws JsonProcessingException, In
+        publishEvents(eventType.getName(), 2, i -> "{\"foo\":\"bar\"}");
         finished.await();
         cursorsDuringPublish = inTimeCursors.get();
+        System.out.println("Total took: " + (System.currentTimeMillis() - start)/1000);
```
do you still need that?

antban replied:
No, will remove

v-stepanov commented on src/main/java/org/zalando/nakadi/ShutdownHooks.java:
```diff
@@ -0,0 +1,50 @@
+            final Runnable hook;
+            synchronized (HOOKS) {
+                hook = HOOKS.isEmpty() ? null : HOOKS.iterator().next();
+                HOOKS.remove(hook);
```
I think `Queue` would be a better data structure for `HOOKS`. Above two lines you would be able to replace with `hook = HOOKS.poll();`

antban replied:
@v-stepanov Operation that is performed here is performed only once and I care more about acting as a set, not as a queue

v-stepanov commented on src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java:
```diff
@@ -134,6 +134,21 @@ private void createTopic(final String topic, final int partitionsNum, final int
+                final List<PartitionInfo> partitions = consumer.partitionsFor(topic);
+                creationConfirmed = null != partitions;
+            }
+        }
```
I think you try to reinvent here `executeWithRetry` which we use in other places.

antban replied:
I thought that we are using it only in tests :(( But from the other hand - no interval, no complex lamdas, and simple.
@v-stepanov I think that it is fine

v-stepanov replied:
1) I don't insist, the code you wrote is fine for me. However I think something like this would be more clear:
```java
final Boolean created = executeWithRetry(() -> {
            try (Consumer<byte[], byte[]> consumer = kafkaFactory.getConsumer()) {
                return consumer.partitionsFor(topic) != null;
            }
        },
        new RetryForSpecifiedTimeStrategy<Boolean>(TimeUnit.SECONDS.toMillis(5))
                .withResultsThatForceRetry(Boolean.FALSE));
if (created != Boolean.TRUE) {
    throw new TopicCreationException("Failed to confirm topic creation within " + timeoutMillis + " millis");
}
```

2) Actually I think that it would be better here to use interval here (and executeWithRetry allows that) to have a control over calling kafka method.

3) Why do you use consumer here not the producer? Creation of producer is not a fast thing.

antban replied:
1. Ok, I will do what you propose. 
2. Added interval of 100 ms
3. Because I am not sure, that behavior of producer and consumer is the same, so I decided to use exactly the same call that was failing.

v-stepanov commented on src/main/java/org/zalando/nakadi/service/subscription/StreamingContext.java:
```diff
@@ -128,9 +129,14 @@ public EventStreamWriter getWriter() {
-        streamInternal(new StartingState());
+        try (Closeable ignore = ShutdownHooks.addHook(this::onNodeShutdown)) { // bugfix ARUHA-485
+            streamInternal(new StartingState());
+        } catch (IOException ex) {
```
final

antban replied:
@v-stepanov will figure out while checkstyle is not checking style...

antban replied:
Failed to find way to configure checkstyle...

v-stepanov commented on src/test/java/org/zalando/nakadi/utils/TestUtils.java:
```diff
@@ -179,11 +176,15 @@ public static void waitFor(final Runnable runnable, final long timeoutMs) {
+                               final Class<? extends  Throwable>... additionalException) {
+        final Class<? extends Throwable>[] leadToRetry = new Class[additionalException.length + 1];
+        System.arraycopy(additionalException, 0, leadToRetry, 1, additionalException.length);
+        leadToRetry[0] = AssertionError.class;
```
You are array lover :)
I think it can be done much easier:
```java
final List<Class<? extends Throwable>> exs = Arrays.asList(additionalException);
exs.add(AssertionError.class);
```

antban replied:
@v-stepanov I do not remember myself writing this code

antban replied:
@v-stepanov But probably i wrote it

v-stepanov commented:
I finished review. In general it's fine for me, I wrote comments regarding some of the issues.
For me it seems that most of the effort in this PR was done towards possibility to run tests in parallel or on "dirty" DB. This is very cool, however I'm not sure of how much it contributes to stability of tests. 
So in general I'm fine with merging of this PR (after comments are addressed) as it brings a lot of valuable stuff, however I don't think that the initial ticket can be resolved.

antban commented:
@v-stepanov Added commit that probably addresses all the comments (https://github.com/zalando/nakadi/pull/709/commits/58d2987b0385385063d9f0524bc72179699483fd)
Regarding wording (indeterministic) - for me it means, that tests are correctly written and can run for many times on the same environment without restarts. For now i believe, that each problem that is found by tests is a real problem that should be investigated and fixed.

v-stepanov commented on build.gradle:
```diff
@@ -315,8 +315,7 @@ task fullAcceptanceTest(type: GradleBuild) {
     classpath = sourceSets.acceptanceTest.runtimeClasspath
-//    maxParallelForks = Runtime.runtime.availableProcessors()
-    maxParallelForks = 1
+    maxParallelForks = Runtime.runtime.availableProcessors()
```
@antban Why have you changed it now? I thought you told that it's not stable yet to run tests in parallel.

antban replied:
I decided to get rid of commented line (cause I do not like commented lines) and for me it works pretty stable and much faster then single-threaded version. 
For travis it didn't change anything.

v-stepanov commented:
I will approve. Also, I think this ticket requires a normal release process as it touches not only tests code but also the main code.

v-stepanov commented:
👍

antban commented:
:+1:

antban commented:
:+1:

v-stepanov commented:
👍

antban commented:
:+1:

lmontrieux commented:
👍
